### PR TITLE
[WIP][SQL] query_builder helper

### PR DESCRIPTION
**Pull request by zach2good**
Behold, some of the worst code I've written in many years!

Until someone (me) gets the gall to tear apart everything to do with how we use the db, a little helper to make adding/removing/changing fields in our raw queries is probably welcome. **The main goal here is to stop us from having to track the raw indexes into SQL results.**

**TODO**
- More and better type checking
- Storing type (and blob/string/container) length at field declaration time

Turns:
```cpp
        int32 ret = Sql_Query(SqlHandle, "SELECT resultId, seed, element1, element2, result, min_quantity, max_quantity, weight FROM gardening_results");

        if (ret != SQL_ERROR && Sql_NumRows(SqlHandle) != 0)
        {
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                uint8 SeedID   = (uint8)Sql_GetUIntData(SqlHandle, 1);
                uint8 Element1 = (uint8)Sql_GetUIntData(SqlHandle, 2);
                uint8 Element2 = (uint8)Sql_GetUIntData(SqlHandle, 3);

                uint32 uid = (SeedID << 8) + (Element1 << 4) + Element2;

                GardenResultList_t& resultList = g_pGardenResultMap[uid];

                uint16 ItemID      = (uint16)Sql_GetIntData(SqlHandle, 4);
                uint8  MinQuantity = (uint8)Sql_GetIntData(SqlHandle, 5);
                uint8  MaxQuantity = (uint8)Sql_GetIntData(SqlHandle, 6);
                uint8  Weight      = (uint8)Sql_GetIntData(SqlHandle, 7);
                resultList.emplace_back(ItemID, MinQuantity, MaxQuantity, Weight);
            }
        }
```
into:
```cpp
        auto qb = query::builder(SqlHandle);
        qb.select()
            .field<uint32>("resultId")
            .field<uint32>("seed")
            .field<uint32>("element1")
            .field<uint32>("element2")
            .field<uint32>("result")
            .field<uint32>("min_quantity")
            .field<uint32>("max_quantity")
            .field<uint32>("weight")
            .from("gardening_results");

        auto results = qb.execute();
        if (!results.m_rows.empty())
        {
            for (auto& entry : results.m_rows)
            {
                uint8 SeedID   = entry.get<uint32>("seed");
                uint8 Element1 = entry.get<uint32>("element1");
                uint8 Element2 = entry.get<uint32>("element1");

                uint32 uid = (SeedID << 8) + (Element1 << 4) + Element2;
                
                GardenResultList_t& resultList = g_pGardenResultMap[uid];
                
                uint16 ItemID      = entry.get<uint32>("result");
                uint8  MinQuantity = entry.get<uint32>("min_quantity");
                uint8  MaxQuantity = entry.get<uint32>("max_quantity");
                uint8  Weight      = entry.get<uint32>("weight");
                
                resultList.emplace_back(ItemID, MinQuantity, MaxQuantity, Weight);
            }
        }
```

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

**Comment by TeoTwawki** [comment]
> [3:16 PM] TeoTwawki: do it and I'll build you a SHRINE
[3:17 PM] TeoTwawki:  quotable

Your shrine sir: ⛩️

**Comment by Jahana** [comment]
Is line78 of src/map/utils/gardenutils.cpp (_**uint8 Element2 = entry.get<uint32>("element1");**_) supposed to be element1 or element2?

**Comment by zach2good** [comment]
> Is line78 of src/map/utils/gardenutils.cpp (uint8 Element2 = entry.get("element1");) supposed to be element1 or element2?

Nope, thats a typo, thanks!

File: src/common/query_builder.h [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+            }
+
+            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
+            {
+                row_t row;
+                for (int i = 0; i < fields.size(); ++i)
+                {
+                    auto name = fields[i].first;
+                    auto type = fields[i].second;
+                    switch (type)
```
Comment (zach2good) [inline src/common/query_builder.h:330]: This can be replaced with `std::visit`, which is constexpr

**Comment by zach2good** [comment]
Going to replace this work with: https://github.com/rbock/sqlpp11

**Comment by TeoTwawki** [comment]
So do we let him keep the shrine boys or no?

**Pull request closed without merging**